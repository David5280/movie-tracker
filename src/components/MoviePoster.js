import React from 'react';
// import { connect } from 'react-redux'
// import { addNewFavorite } from '../utilz/apiCalls'
// import { addFavorite, viewFavorites, deleteFavorite } from '../actions'
import activeFavIcon from '../images/001-heart.svg';
// import { statements } from '@babel/template';
// import { compose } from '../../../../../Library/Caches/typescript/3.4.3/node_modules/redux';
// import inactiveFavIcon from '../images/002-heart-1.svg'

class MoviePoster extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      isFavored: false
      // currentHover: false
    };
  }

  setHover = e => {
    e.preventDefault();
    // this.setState({ currentHover: !this.state.currentHover });
    this.props.findMovie(parseInt(e.target.id));
  };

  cancelFocus = () => {
    this.props.cancelFocus();
  };

  render() {
    return (
      <article className="movie-poster">
        <img
          src={`https://image.tmdb.org/t/p/w500/${this.props.posterPath}`}
          alt={`${this.props.title}-poster`}
          className="poster-img"
          id={`${this.props.id}`}
          onMouseEnter={e => this.setHover(e)}
          onMouseLeave={() => this.cancelFocus()}
        />
        <p className="poster-title">{this.props.title}</p>
        <img
          src={activeFavIcon}
          className="favorite-icon"
          alt="favorite-icon"
        />
      </article>
    );
  }
}

export default MoviePoster;
