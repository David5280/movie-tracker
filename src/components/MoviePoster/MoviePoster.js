import React from 'react';
import { connect } from 'react-redux';
import { withRouter, NavLink } from 'react-router-dom';
import { loadFavorites } from '../../actions';
import inactiveFave from '../../images/012-hollywood-star.svg';
import activeFave from '../../images/011-cinema.svg';
import PropTypes from 'prop-types';
import {
  postFavorite,
  removeFavorite,
  fetchFavorites
} from '../../utilz/apiCalls';

export class MoviePoster extends React.Component {
  setHover = e => {
    e.preventDefault();
    this.props.findMovie(parseInt(e.target.closest('article').id));
  };

  cancelFocus = () => {
    this.props.cancelFocus();
  };

  handleFavorite = event => {
    event.preventDefault();
    const { user, movie, history } = this.props;
    if (user.id) {
      this.toggleFavorite(user.id, movie.movie_id);
    } else {
      history.push('/login');
    }
  };

  toggleFavorite(userID, movieID) {
    const { favorites } = this.props;
    const favoriteIDs = favorites.map(movie => movie.movie_id);
    if (!favoriteIDs.includes(movieID)) {
      this.addFavorite(userID);
      this.setState({ isFavored: true });
    } else {
      this.deleteFavorite(userID, movieID);
      this.setState({ isFavored: false });
    }
  }

  addFavorite = userID => {
    const { movie } = this.props;
    postFavorite(userID, movie).then(() =>
      fetchFavorites(userID).then(movies => this.props.loadFavorites(movies))
    );
  };

  deleteFavorite = (userID, movieID) => {
    removeFavorite(userID, movieID).then(() =>
      fetchFavorites(userID).then(movies => this.props.loadFavorites(movies))
    );
  };

  determineFavorite = () => {
    const { favorites, movie } = this.props;
    if (!favorites) {
      return inactiveFave;
    } else {
      const isFavorite = favorites.find(
        fave => fave.movie_id === movie.movie_id
      );
      return isFavorite ? activeFave : inactiveFave;
    }
  };

  render() {
    const { movie } = this.props;
    return (
      <article
        className="movie-poster"
        id={movie.movie_id}
        onMouseLeave={() => this.cancelFocus()}
      >
        <img
          src={this.determineFavorite()}
          className="favorite-icon"
          alt="fav-icon"
          onClick={e => this.handleFavorite(e)}
        />
        <NavLink
          to={`/movies/${movie.movie_id}`}
          className="movie-showcase-link"
          onMouseEnter={e => this.setHover(e)}
        >
          <img
            src={`https://image.tmdb.org/t/p/w1280/${movie.poster_path}`}
            alt={`${movie.title}-poster`}
            className="poster-img"
            onMouseEnter={e => this.setHover(e)}
          />
        </NavLink>
        <p className="poster-title">{movie.title}</p>
      </article>
    );
  }
}

const mapStateToProps = state => ({
  favorites: state.favorites,
  user: state.user
});

const mapDispatchToProps = dispatch => ({
  loadFavorites: movies => dispatch(loadFavorites(movies))
});

MoviePoster.propTypes = {
  movie: PropTypes.object.isRequired
};

export default withRouter(
  connect(
    mapStateToProps,
    mapDispatchToProps
  )(MoviePoster)
);
