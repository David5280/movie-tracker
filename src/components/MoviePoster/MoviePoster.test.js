import React from 'react';
import { shallow } from 'enzyme';
import { MoviePoster } from './MoviePoster';

describe('MoviePoster', () => {
  let wrapper;
  let props;
  let mockEvent;
  let instance;

  beforeEach(() => {
    props = {
      user: { id: 1 },
      findMovie: jest.fn(),
      cancelFocus: jest.fn(),
      movie: { movie_id: 123, title: 'yeet', poster_path: 'path' }
    };
    mockEvent = {
      preventDefault: () => jest.fn(),
      target: {
        closest: () => ({ id: 1 })
      }
    };
    wrapper = shallow(<MoviePoster {...props} />);
    instance = wrapper.instance();
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should be able to find movie on hover', () => {
    instance.setHover = jest.fn();
    wrapper.find('.poster-img').simulate('mouseEnter', mockEvent);
    expect(instance.setHover).toHaveBeenCalled();
  });

  it('should be able to cancel focus', () => {
    instance.cancelFocus = jest.fn();
    wrapper.find('.movie-poster').simulate('mouseLeave');
    expect(instance.cancelFocus).toHaveBeenCalled();
  });

  it('should only allow users to mark favorites', () => {
    instance.handleFavorite = jest.fn();
    wrapper.find('.favorite-icon').simulate('click');
    expect(instance.handleFavorite).toHaveBeenCalled();
  });
});
