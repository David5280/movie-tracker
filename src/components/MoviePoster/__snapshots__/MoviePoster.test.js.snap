// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MoviePoster should match snapshot 1`] = `ShallowWrapper {}`;
